doc/ractor.md

```diff
@@ -0,0 +1,882 @@
+
+Only main Ractor (a Ractor created at starting of interpreter) can access global variables.
+
+```ruby
+  $gv = 1
+  r = Ractor.new do
+    $gv
+  end
+
+  begin
+    r.take
+  rescue Ractor::RemoteError => e
+    e.cause.message #=> 'can not access global variables from non-main Ractors'
+  end
+```
+
+### Instance variables of shareable objects
+
+Only main Ractor can access instance variables of shareable objects.
+
+```ruby
+  class C
+    @iv = 'str'
+  end
+
+  r = Ractor.new do
+    class C
+      p @iv
+    end
+  end
+
+
+  begin
+    r.take
+  rescue => e
+    e.class #=> RuntimeError
+  end
+```
+
+```ruby
+  shared = Ractor.new{}
+  shared.instance_variable_set(:@iv, 'str')
+
+  r = Ractor.new shared do |shared|
+    p shared.instance_variable_get(:@iv)
+  end
+
+  begin
+    r.take
+  rescue Ractor::RemoteError => e
+    e.cause.message
+  end
+```
+
+Note that instance variables for class/module objects are also prohibited on Ractors.
+
+### Class variables
+
+Only main Ractor can access class variables.
+
+```ruby
+  class C
+    @@cv = 'str'
+  end
+
+  r = Ractor.new do
+    class C
+      p @@cv
+    end
+  end
+
+
+  begin
+    r.take
+  rescue => e
+    e.class #=> RuntimeError
+  end
+```
+
+### Constants
+
+Only main Ractor can read constants which refer to the unshareable object.
+
+```ruby
+  class C
+    CONST = 'str'
+  end
+  r = Ractor.new do
+    C::CONST
+  end
+  begin
+    r.take
+  rescue => e
+    e.class #=> NameError
+  end
+```
+
+Only main Ractor can define constants which refer to the unshareable object.
+
+```ruby
+  class C
+  end
+  r = Ractor.new do
+    C::CONST = 'str'
+  end
+  begin
+    r.take
+  rescue => e
+    e.class
+  end
+```
+
+## Implementation note
+
+* Each Ractor has own thread, it means each Ractor has at least 1 native thread.
+* Each Ractor has own ID (`rb_ractor_t::id`).
+  * On debug mode, all unshareable objects are labeled with current Ractor's id, and it is checked to detect unshareable object leak (access an object from different Ractor) in VM.
+
+## Examples
+
+### Traditional Ring example in Actor-model
+
+```ruby
+RN = 10000
+CR = Ractor.current
+
+last_r = r = Ractor.new do
+  p Ractor.recv
+  CR << :fin
+end
+
+RN.times{
+  r = Ractor.new r do |next_r|
+    next_r << Ractor.recv
+  end
+}
+
+p :setup_ok
+r << 1
+p Ractor.recv
+```
+
+### Fork-join
+
+```ruby
+def fib n
+  if n < 2
+    1
+  else
+    fib(n-2) + fib(n-1)
+  end
+end
+
+RN = 10
+rs = (1..RN).map do |i|
+  Ractor.new i do |i|
+    [i, fib(i)]
+  end
+end
+
+until rs.empty?
+  r, v = Ractor.select(*rs)
+  rs.delete r
+  p answer: v
+end
+```
+
+### Worker pool
+
+```ruby
+require 'prime'
+
+pipe = Ractor.new do
+  loop do
+    Ractor.yield Ractor.recv
+  end
+end
+
+N = 1000
+RN = 10
+workers = (1..RN).map do
+  Ractor.new pipe do |pipe|
+    while n = pipe.take
+      Ractor.yield [n, n.prime?]
+    end
+  end
+end
+
+(1..N).each{|i|
+  pipe << i
+}
+
+pp (1..N).map{
+  r, (n, b) = Ractor.select(*workers)
```

Comment:
Maybe use `_r` to mention that `r` is never used?

```suggestion
  _r, (n, b) = Ractor.select(*workers)
```

Reply:
good idea.